### ISSUE TITLE
Content stream operators are "PDF objects"?

### ISSUE THREAD

**Issue by pesco** [pdf-association/pdf-issues#201]
**Describe the bug**

There is an internal inconsistency in the wording of Section 7.8.2 "Content streams". Its paragraph 1 makes the following statement:

> A content stream ... consists of a sequence of instructions ...
> The instructions shall be represented in the form of PDF objects,
> using the same object syntax as in the rest of the PDF document.

It seems clear that operators are not PDF objects. However, paragraph 3 reiterates:

> [A content stream] consists of PDF objects denoting operands and operators.

This can be read in two ways: (a) "PDF objects denoting operators and PDF objects denoting operators" or (b) "PDF objects denoting operators and other things called operators". I am assuming that the term "PDF object" in the context of this specification can only reasonably be taken to refer to objects as defined by section 7.3.

Paragraph 5 clarifies:

> An operator is a PDF keyword ...
> [They] shall be distinguished from a name object by the
> absence of an initial SOLIDUS character (2Fh) (/).
> Operators shall be meaningful only inside a content stream.

The term "PDF keyword" is not specifically defined, so it presumably means the same thing as the generic term "keyword" as used, for example, in section 7.3 "Objects".

**Recommendation for correction**

Replace the offending sentence in paragraph 1 as follows:
```diff
- The instructions shall be represented in the form of PDF objects, using the same object syntax as in the rest of the PDF document.
+ The instructions make use of PDF objects which shall be represented using the same syntax as in the rest of the PDF document.
```

Reword the sentence in paragraph 3 as follows:
```diff
- It consists of PDF objects denoting operands and operators.
+ It consists of PDF objects denoting operands and keywords denoting operators.
```

Remove the word "PDF" from the term "keyword" in paragraph 5:
```diff
- An operator is a PDF keyword specifying some action
+ An operator is a keyword specifying some action
```

**Comment by lrosenthol** [comment]
I am OK with the first proposed changed and the last, but not the middle one, as (was discussed in the Securing PDF TWG) operators are also PDF objects.

**Comment by pesco** [comment]
Hi, thanks for discussing this! Could you elaborate why/how PDF operators are "PDF objects"? To quote myself:

> I am assuming that the term "PDF object" in the context of this specification can only reasonably be taken to refer to objects as defined by section 7.3.

Section 7.3, of course, does not define operators. This is the core of the perceived inconsistency that prompted me to open this issue.

**Comment by lrosenthol** [comment]
Yes, that is correct - operators are PDF objects.  It states this clearly in 7.8.2 (paragraph 3)

>A content stream, after decoding with any specified filters, shall be interpreted according to the PDF syntax rules described in 7.2, "Lexical conventions". It consists of PDF objects denoting operands and operators.

Two paragraphs after that (p5) it clearly states what an operator is:

>An operator is a PDF keyword specifying some action that shall be performed, such as painting a graphical shape on the page. An operator keyword shall be distinguished from a name object by the absence of an initial SOLIDUS character (2Fh) (/). Operators shall be meaningful only inside a content stream.

Pretty clear.

However, as discussed here and in other threads, I think we need to fix the use of keyword here...

**Comment by petervwyatt** [comment]
I will propose editorial changes for review in a future PDF TWG meeting.

**Comment by pesco** [comment]
> Yes, that is correct - operators are PDF objects. It states this clearly in 7.8.2 (paragraph 3)
> 
> > A content stream ... consists of PDF objects denoting operands and operators.

I don't know, I think reading that sentence as a definition of objects is *very* backwards.

**Comment by lrosenthol** [comment]
>I think reading that sentence as a definition of objects is very backwards.

It is **NOT** a definition of objects, but is a **USE** of objects - just as the text reads.

**Comment by lrosenthol** [comment]
Definition of what @pesco ?   

Operators and Operands in a content stream are PDF objects.  PDF objects are those things defined in section 7.3, Objects.

7.8.2 goes even further in clarifying the types of objects for each:
>An operand is a direct object belonging to any of the basic PDF data types except a stream. Dictionaries shall be permitted as operands only by certain specific operators. Indirect objects and object references shall not be permitted at all.
An operator is a PDF keyword specifying some action that shall be performed, such as painting a graphical shape on the page. An operator keyword shall be distinguished from a name object by the absence of an initial SOLIDUS character (2Fh) (/). Operators shall be meaningful only inside a content stream.

Do you think we need more clarification than all of that?

**Comment by pesco** [comment]
> PDF objects are those things defined in section 7.3, Objects.

> Section 7.3, of course, does not define operators. 

How much more clarification do you need, Mr. Rosenthol?

**Comment by lrosenthol** [comment]
Oh - I think you are saying that you believe that they should all be defined in a single section of the specification.  Yes?

I don't see any reason that needs to be the case.   Things are defined where they need to be defined.  In this case, you don't need to worry about operators and operands until you start worrying about Content Streams - and so when you do, you now get additional definitions that are relevant.

**Comment by petervwyatt** [comment]
This and other related/interlinked issues (#194 #199 #202 #208 #209) are being discussed in the "Securing PDF" DG of ISO TC 171 SC 2 WG 8 and will be labelled as "Parked" here in GitHub until such time as a set of solutions can be proposed.

**Comment by crlf0710** [comment]
Might or might not be related: I always felt the "Terms and definitions" clause is kind of lacking, PDF technology introduced so many terms, while relatively a large portion is absent from the terms and definitions clause.

**Comment by petervwyatt** [comment]
Errata #363 proposes to formally define "PDF keyword" as a lexical construct which I think flows through and also solves this issue.

**Comment by petervwyatt** [comment]
Thanks @pesco - a "PDF keyword" as a lexical convention means it is not an object so that sentence needs slight rewording. I also note that the binary data of an inline image is also neither! This is a descriptive sentence and not a requirement (no "shall") so we can take some liberty:

> [A content stream] **primarily** consists of PDF objects **and PDF keywords** denoting operands and operators.

**Comment by petervwyatt** [comment]
Will be resolved as part of the resolution to Errata #363.

**Comment by car222222** [comment]
So let us hope that the outcome of #363 does finally make it clear crystal clear (in all the necessary places) that, despite the repeated protestations of @lrosenthol , PDF operators are *no longer*, in any cases, PDF objects. 

(If they ever were! :-)
